B 15 add new order

shaguahill opened the pull request:
[b-15] Add new Order:

1. Created sql request and method for insert new order to DB
2. Added link /order to html files
3. Added method post for add new order
4. Changed new-order.html
5. Fix bug with get photo from empty list
6. Resolve conflicts
7. Add path to WebServlet after merge from develop

shaguahill committed:
[b-15] Add new Order:
1. Created sql request and method for insert new order to DB
2. Added link /order to html files
3. Added method post for add new order
4. Changed new-order.html
5. Fix bug with get photo from empty list

shaguahill committed:
Merge branch 'develop' into b-15-add-new-order

# Conflicts:
#	src/main/java/com/photostudio/service/impl/DefaultOrderService.java
#	src/main/java/com/photostudio/web/servlet/order/OrderServlet.java
#	src/main/webapp/WEB-INF/web.xml

Resolve conflicts

shaguahill committed:
Merge branch 'develop' into b-15-add-new-order
Add path to WebServlet

viktorterljuk commented on src/main/java/com/photostudio/dao/jdbc/JdbcOrderDao.java:
```diff
@@ -180,6 +183,22 @@ public void delete(long id) {
+            preparedStatement.setString(2, order.getUser().getEmail());
+            preparedStatement.setString(3, order.getComment());
+            preparedStatement.executeUpdate();
+            LOG.info("Order created and added to DB");
```
add order object to the log message

shaguahill replied:
Fixed

shaguahill committed:
b-15-add-new-order(Fixed bugs after review)
Added object order to logs

Rooman commented on src/main/java/com/photostudio/web/servlet/order/OrderServlet.java:
```diff
@@ -14,39 +15,94 @@
 import java.util.Map;
 
-@WebServlet(urlPatterns = "/order/*")
+@WebServlet(urlPatterns = {"/order", "/order/*"})
```
разнести в два сервлета

shaguahill replied:
Разнес

Rooman commented on src/main/java/com/photostudio/web/servlet/order/OrderServlet.java:
```diff
@@ -14,39 +15,94 @@
+                if (OrderStatus.NEW.equals(order.getStatus())) {
+                    paramsMap.put("order", order);
+                    response.setContentType("text/html;charset=utf-8");
+                    TemplateEngineFactory.process(request, response, "new-order", paramsMap);
```
мне кажется что если это запрос на GET /order/{id} но это не new order.Это просто order. Название шаблона сбивает.

shaguahill replied:
Название изменил

shaguahill committed:
b-15-add-new-order(Fixed bugs after review)
1. Moved functional create new order to another servlet
2. Changed name html file from "new-order" to "order"
3. Removed block "form" from order.html and created script for button "Save" with post request

Rooman commented on src/main/java/com/photostudio/dao/jdbc/JdbcOrderDao.java:
```diff
@@ -30,11 +30,14 @@
     private static final String DELETE_ORDER_BY_ID = "DELETE FROM Orders WHERE id = ?";
 
+    private static final String ADD_NEW_ORDER = "INSERT INTO Orders (orderDate, statusId, userId, comment) VALUES (?, " +
+            "(SELECT id FROM OrderStatus WHERE statusName='New'),(SELECT id FROM Users WHERE email=?), ?)";
```
OrderStatus <- статические данные которые не будут меняться. Мне не нравится что мы делаем вложенный селект. Можешь закешировать эти значения?
Зачем вложенный селект на юзеров? Если у нас есть имейл, почему у нас нету айдишника?

shaguahill replied:
Done

Rooman commented on src/main/java/com/photostudio/web/servlet/order/AddNewOrderServlet.java:
```diff
@@ -0,0 +1,65 @@
+
+        String comment = request.getParameter("comment");
+
+        Order.OrderBuilder order = Order.builder().orderDate(LocalDateTime.now())
```
это orderBuilder

shaguahill replied:
Renamed

Rooman commented on src/main/java/com/photostudio/web/servlet/order/AddNewOrderServlet.java:
```diff
@@ -0,0 +1,65 @@
+                .status(OrderStatus.NEW)
+                .user(user);
+
+        if (comment.equals("")) {
```
в три строки пожалуйста этот код

shaguahill replied:
Done

Rooman commented on src/main/java/com/photostudio/web/servlet/order/OrderServlet.java:
```diff
@@ -25,28 +27,35 @@
         }
     }
+
+    private String getLastPartOfUri(HttpServletRequest request) {
```
Если бы ты тестировал этот метод, что стоило бы сделать, то ты бы передавал сюда не request а String uri. 
И название метода getIdFromURI потому что по факту он достает конкретно идишник в твоей конкретной задаче.

shaguahill replied:
Done

Rooman commented on src/main/webapp/WEB-INF/templates/admin-page.html:
```diff
@@ -18,7 +18,7 @@ <h1 class="admin__title">Admin page</h1>
             <ul class="admin__nav-links">
                 <li class="admin__nav-link">
-                    <a href="#" class="btn btn-submit btn-outline-dark button">New order</a>
+                    <a href="order" class="btn btn-submit btn-outline-dark button">New order</a>
```
это точно будет работать? Это относительная линка? Не нужно ли делать th:href?

shaguahill replied:
Так как у нас пока что запросы на эту страницу идут с корневой папки  - это работает. Но на всякий случай, если что-то поменяется - изменил на th:href

Rooman commented on src/main/webapp/WEB-INF/templates/order.html:
```diff
@@ -36,23 +36,32 @@ <h1 class="order__title">Order</h1>
+            <textarea th:if="${order.comment} != null" class="form-control order__detail-info order__comment"
+                      id="comment-exist" rows="5" name="comment"
+                      th:text="${order.comment}"></textarea>
+            <textarea th:if="${order.comment} == null" class="form-control order__detail-info order__comment"
```
объясни пожалуйста зачем это разбивать на две части. Нельзя ли объеделить и th:text и placehlder

shaguahill replied:
Объединил

Rooman commented on src/main/webapp/assets/js/order/new-order.js:
```diff
@@ -0,0 +1,19 @@
+    if (email === "") {
+        alert("Email field empty");
+    } else {
+        $.post({
```
@Brutt Юра, мы используем джаваскрипт или нет?

Brutt replied:
В нас є деяка частина функціонала на JS, яка стосується юзерів.

Rooman replied:
@Brutt В данном контексте как поступает? Мы используем JS для нового кода?

Brutt replied:
Не використовуємо JS для нового коду

shaguahill replied:
Изменил на form. Но так как вложеных форм не бывает, проблему это не решило.

Rooman reviewed (changes requested):
Можно ли в конектсте данной задачи внести изменения по скриптам для базы которые мы обсуждали вчера?

Rooman commented on src/main/java/com/photostudio/dao/jdbc/mapper/OrderWithPhotoRowMapper.java:
```diff
@@ -30,7 +30,10 @@ public Order mapRow(ResultSet resultSet) throws SQLException {
             }
-            photoSources.add(resultSet.getString("source"));
+            String source = resultSet.getString("source");
+            if (source != null) {
```
меня смущает эта строка... 
Тут точно только так можно написать?

shaguahill replied:
Не нашел другого способа. Как будет лучше?

shaguahill committed:
b-15-add-new-order(Fixed bugs after review)
1. User flyway for test:
 1.1. Deleted command COLLATE from V1_0_initial_schema
 1.2. Added Constraint name for foreign key
 1.3. Separated requests in V1_4_*
 1.4. Changed TestDataSource and small changes in test files.
2. Deleted nested SELECTS
3. Created Cache for OrderStatus
4. Deleted js files for button save in order.html, and rewrote this file used <form>

shaguahill committed:
Merge branch 'develop' into b-15-add-new-order

# Conflicts:
#	src/main/webapp/WEB-INF/templates/all-orders.html
#	src/main/webapp/WEB-INF/templates/fragment/navigation.html
#	src/main/webapp/WEB-INF/templates/new-order.html

Conflicted resolved

shaguahill committed:
b-15-add-new-order(Fix bug):
1. Added condition for form for add new order

shaguahill committed:
b-15-add-new-order(Test Fix):
